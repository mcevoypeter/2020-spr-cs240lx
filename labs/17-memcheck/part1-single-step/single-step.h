#ifndef __SINGLE_STEP_H__
#define __SINGLE_STEP_H__

// initialize single step mode.
int single_step_init(void);

// run a single routine <user_fn> in single step mode using stack <sp>
int single_step_run(int user_fn(void), uint32_t sp);

#endif
