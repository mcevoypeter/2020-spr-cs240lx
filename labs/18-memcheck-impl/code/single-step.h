#ifndef __SINGLE_STEP_H__
#define __SINGLE_STEP_H__

#include "rpi.h"
#include "cp14-debug.h"
#include "bit-support.h"

// initialize single step mode: 1 on success.
int single_step_init(void);

// run a single routine <user_fn> in single step mode using stack <sp>
int single_step_run(int user_fn(void), uint32_t sp);

// return the number of instructions we single stepped.
//  (just count the faults)
unsigned single_step_cnt(void);

#endif
