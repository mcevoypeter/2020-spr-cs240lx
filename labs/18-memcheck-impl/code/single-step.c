#include "rpi.h"
#include "rpi-interrupts.h"
#include "single-step.h"
#include "cp14-debug.h"
#include "libc/helper-macros.h"
#include "sys-call-asm.h"

int single_step_init(void) {
    cp14_enable();
    return 1;
}

static unsigned step_cnt = 0;
static void single_step_handler(uint32_t regs[16], uint32_t pc, uint32_t addr) {
    step_cnt++; 
    brkpt_mismatch_set0(pc, single_step_handler);    
}

int single_step_run(int user_fn(void), uint32_t sp) {
    step_cnt = 0;

    brkpt_mismatch_set0(0, single_step_handler);

    int ret_val = user_mode_run_fn(user_fn, sp);

    brkpt_mismatch_disable0(0);

    return ret_val;
}

unsigned single_step_cnt(void) {
    return step_cnt;
}
