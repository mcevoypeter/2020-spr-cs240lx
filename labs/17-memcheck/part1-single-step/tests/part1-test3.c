// engler: trivial check that you can switch to user mode and back.
#include "single-step.h"

static int user_fn(void) {
    asm volatile("nop");  // 1
    asm volatile("nop");  // 2
    asm volatile("nop");  // 3
    asm volatile("nop");  // 4
    asm volatile("nop");  // 5

    asm volatile("nop");  // 1
    asm volatile("nop");  // 2
    asm volatile("nop");  // 3
    asm volatile("nop");  // 4
    asm volatile("nop");  // 5

    return 0xdeadbeef;
}

void notmain() {
    single_step_init();

    assert(mode_is_super());
    trace("about to try running at user level\n");
    unsigned x = single_step_run(user_fn, STACK_ADDR2);
    assert(x == 0xdeadbeef);
    assert(mode_is_super());
    trace("success: %d single steps (i got 17)!\n", single_step_cnt());
}
