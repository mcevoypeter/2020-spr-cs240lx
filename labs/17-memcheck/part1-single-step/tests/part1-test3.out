find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./tests/part1-test3.bin> about to boot
******************sending 9920 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:about to try running at user level
cur mode: SUPER MODE, saved mode = USER MODE, switch-to mode=SUPER MODE
shadow lr=0x8f40, shadow sp = 0x7000000
pc = 0x8f48, saved sp=0x7000000
system call: sp=0x7ffff78, inst=11101111000000000000000000000001, sys_num=1, r0=0x60000113
TRACE:success: 17 single steps (i got 17)!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
